### ISSUE TITLE
Feature Request: Allow synchronizing smallest files first.

### ISSUE THREAD

**Issue by RokeJulianLockhart** [abraunegg/onedrive#2544]
### Is your feature request related to a problem? Please describe.

I've a very slow internet connection, and frequently need to move from place to place. This means that I need to be able to sync what I can in the timeframe I am provided.

### Describe the solution you'd like

In my case, I solely care about being able to synchronize the smallest files first.

### Describe alternatives you've considered

Allow the user to order how the files from the server should be synchronized more generally, including this. This is the best course of action because it means the logic is less custom for each option added subsequently.

### Additional context

> https://github.com/abraunegg/onedrive/discussions/2543#discussioncomment-7594337

**Comment by abraunegg** [comment]
@RokeJulianLockhart , @Are52Tap , @ecogit

Please can you provide some further details here:

> In my case, I solely care about being able to synchronize the smallest files first.

* Download only ?
* Upload only ?
* Both

**Comment by RokeJulianLockhart** [comment]
> https://github.com/abraunegg/onedrive/issues/2544#issuecomment-2649167281

@abraunegg, I was thinking of download when I requested this, but both would be more sensical. Do you have a preference?

**Comment by abraunegg** [comment]
> > [#2544 (comment)](https://github.com/abraunegg/onedrive/issues/2544#issuecomment-2649167281)
> 
> [@abraunegg](https://github.com/abraunegg), I was thinking of download when I requested this, but both would be more sensical. Do you have a preference?

No preference. 

I was going to do both, however upload sorting is a little more CPU intensive based on how 'upload' flags what has changed (new,modified).

**Comment by abraunegg** [comment]
> > [#2544 (comment)](https://github.com/abraunegg/onedrive/issues/2544#issuecomment-2649237215)
> 
> [@abraunegg](https://github.com/abraunegg), would allowing the user to choose be feasible?

More options = more chance for bugs.

The hit is small - most likely not even noticable. As the array storing local files that have changed is a pathname - that needs to be queried to obtain size, to sort via size.

The alternative is to change how that array operates to include that datab, but the query to obtain the size still has to be done - so there is no trade off / better mechanism for uploads.

Downloads are different and already implemented in my local pr branch for this feature request.

**Comment by abraunegg** [comment]
@RokeJulianLockhart , @Are52Tap , @ecogit

Please can you test the PR below that implements your feature request.

First install all the require platform dependencies to build the client on your respective platforms. Please read https://github.com/abraunegg/onedrive/blob/master/docs/install.md#building-from-source---high-level-requirements and then follow correctly for your platform.

> [!IMPORTANT]
> Currently there is an issue with installing the DMD or LDC compiler using the documentation instructions:
> `curl -fsS https://dlang.org/install.sh | bash -s dmd` or `curl -fsS https://dlang.org/install.sh | bash -s ldc`
>
> This is not something I can solve and I have emailed the appropriate folk and raised https://github.com/dlang/dmd/issues/20830
> 
> Please augment the installation of DMD or LDC with your distribution version if it is available

Once this is done, to clone the PR to resolve your issue, you can use a script like the following:
```
#!/bin/bash

PR=3114

rm -rf ./onedrive-pr${PR}
git clone https://github.com/abraunegg/onedrive.git onedrive-pr${PR}
cd onedrive-pr${PR}
git fetch origin pull/${PR}/head:pr${PR}
git checkout pr${PR}

# Configure and Build
./configure --enable-debug --enable-notifications; make clean; make;
./onedrive --version
```

This script will create a local folder called `onedrive-pr3114` with the PR version.

To run the PR, you need to run the client from the PR build directory:
```
./onedrive <any other options needed>
```
To install the PR, you will need to perform `sudo make install` to install the PR version to your system.

When running the PR, your version should be: `onedrive v2.5.4-5-gf4ec7c9` or greater.

To enable this new capability, add one of the following to your config file:

* Transfer by file size, smallest first: `transfer_order = "size_asc"`
* Transfer by file size, smallest first: `transfer_order = "size_dsc"`
* Transfer by file name, A -> Z: `transfer_order = "name_asc"`
* Transfer by file name, Z -> A: `transfer_order = "name_dsc"`

**Comment by RokeJulianLockhart** [comment]
> https://github.com/abraunegg/onedrive/issues/2544#issuecomment-2654569002

@abraunegg, you're very kind to provide instructions. However:

1. My OneDrive is utterly full of approximately 2TiB recovery data RN that I can't risk even vaguely corrupting.
2. I can't install unpackaged software onto my primary machine in order to ensure that it remains as reproducible as is possible.

Consequently, I need to:

1. Wait until I download that onto local drives
2. Configure a VM that I can clone the source and compile onto.

Instead, can I defer to you to ascertain whether it's been implemented? I ask because, being the implementer, you probably have a much better idea.

**Comment by RokeJulianLockhart** [comment]
> https://github.com/abraunegg/onedrive/issues/2544#event-16286122539

Ah, sorry. Did you want it closed upon PR merge?

**Comment by abraunegg** [comment]
@RokeJulianLockhart 
This PR was tested during development. It would have been good to get your test to ensure this has been implemented along what you were requesting.

Will mark this as 'implemented' and merge this into 'master'

**Comment by RokeJulianLockhart** [comment]
> https://github.com/abraunegg/onedrive/issues/2544#issuecomment-2655835057

I apologise for not being able to assist.

**Comment by abraunegg** [comment]
This issue has been automatically locked since there has not been any recent activity after it was closed. Please open a new issue for related bugs.
